waspc/data/Generator/templates/server/src/utils.ts

```diff
@@ -0,0 +1,72 @@
+{{={= =}=}}
+import { Request, Response, NextFunction } from 'express'
+import Prisma from '@prisma/client'
+
+import { readdir } from 'fs'
+import { join, dirname } from 'path'
+import { fileURLToPath } from 'url'
+
+{=# isAuthEnabled =}
+import { type SanitizedUser } from './_types/index.js'
+{=/ isAuthEnabled =}
+
+/**
+ * Decorator for async express middleware that handles promise rejections.
+ * @param {Func} middleware - Express middleware function.
+ * @returns {Func} Express middleware that is exactly the same as the given middleware but,
+ *   if given middleware returns promise, reject of that promise will be correctly handled,
+ *   meaning that error will be forwarded to next().
+ */
+type RequestWithExtraFields = Request & {
```

Comment:
> Typescript will anyway allow extra fields (except for inline literal objects), right?

But, we add extra fields if there is an auth token present in our auth middleware. 

I've needed to add these extra fields to the default Express `Request` type to allow me to add `.user` on the `req` in the middleware. (This was needed now since I moved the auth middleware from JS to TS world)


About the name .. I didn't have a better one in mind since in theory it can be any additional fields that we might need in the future. `WaspSpecificRequest`? Or `RequestWithWaspSpecificFields` ... or just `RequestWithExtraFields`?

What do you prefer?

Reply:
I think the current solution works well and it's simple enough. If we have auth enabled, the request will always have an optional `user` on it. If we don't have the auth enabled, the request will not have the extra `user` field on it. 

The name for me is also very clear, it's not just the Express's `Request` it's our Wasp specific Request with extra fields. Hence,`RequestWithExtraFields` but I could go for `WaspRequest` as well 🤷 

Doing the `Request<WithUser>` doesn't make sense to me because when auth is enabled, the `user` field becomes a thing and it's _always_ present, maybe it's value is `undefined`, but that's a runtime thing that needs to be checked. Having `handleRejection` with a type that sometimes lies i.e. `Request` vs. `Request<WithUser>` doesn't feel correct to me.

Reply:
I get that, but what I was saying is that even when auth is enabled, handleRejection will sometimes be dealing with requests that don't have user in them. I provided an example for that for you above. That is because that function is a general utility function, not a function used only for operations/api defined by user. I think you maybe skipped some finer points of my previous comment.

Reply:
To address the example of the code that _doesn't use `user`_, that's the middleware that sets the `user` on the request in the first place.

![Screenshot 2024-01-03 at 13 51 37](https://github.com/wasp-lang/wasp/assets/2223680/d4310322-5722-4623-bcab-4b1a046149c2)

This file is `auth.js` so there would be no Typescript errors, but if we converted this file to Typescript and wanted to write the same piece of code, it would gives us a type error since there isn't a `user` field on the `req` and it can't be set.